### DISCUSSION TITLE
XADD Order on Cluster

### DISCUSSION THREAD

**Discussion by dwlewi4** [redis/lettuce#2207] [Q&A]
Hi

I'm building some integration tests where I am appending batches of messages to a topic with repeated calls to `AbstractRedisAsyncCommands.xadd()` and manually flushing after every command. 

Through logging, I see the cluster is processing 10 - 12 messages a millisecond. 

Although the messages are sent in sequential order (verified using a unique ID in the data), the Stream ID assigned to them are out of order. Intermittently about one and 10 messages ends up very early in the ordering. 

Is this expected behavior for async calls? It's almost like the Stream ID (and thus the order on the stream) isn't assigned until the future is completed?

**Comment by mp911de** [comment]
Calling `xadd(…)` translates to `XADD <key> * <field1> <value1>`. Lettuce uses the `<key>` to calculate the slot hash and the target node. Manual flushing requires that all cluster connections are established, otherwise it can happen that you get a future that creates a connection first, then, because of timing you flush and only after connection creation, the command remains unflushed on the command queue because the flush happened before the command was able to make it to the connection.
